### PR TITLE
[tests] Add test for environment variables and auxiliary vector

### PR DESCRIPTION
**Pull request by oshogbo**
## Description of the changes 
This change adds tests for verifying the environment variables and auxiliary vector.
Those additional tests were requested during the review of PR https://github.com/gramineproject/gramine/pull/541.

The environment variables test is straightforward. The only difference is a custom manifest file, which
sets up the environment variables. Please notice that we care about the order in this test, which isn't required by the ABI documentation. Removing this requirement would be over complication without any additional value.

In the case of the auxiliary vector, the ABI document doesn't specify which types are required.
I decided that the best approach would be to strictly verify the types that Gramine is setting up at this time.

- Lack of any auxiliary entry will cause the test to fail.
- Duplicate entries will cause the test to fail.
- Any additional entries will _NOT_ cause the test to fail.

I decided that the nine entries we have at this point are the minimum required for Gramine to work.
One of the reasons for the missing entry might be that the stack was corrupted.
Another situation of missing entry is that the Gramine developer has changed the vector, which means that the test requires an update. I thought about strict verification of the auxiliary vector. As a strict verification I mean don't allow allow any additional values. But I haven't reached any decisive conclusion.

Please take a moment also to consider the choice of data structure in this test. At this point, we are using an array that behaves like a list. I was thinking about converting this to be an indexed array where we can verify existence using indexing `exists[AT_TYPE]`. The issue with this is that the numbers of the auxiliary vector are split, and an actual number of them is unknown. At this point, I decided that making a list was a good enough approach because the list itself is quite small.

The next step for these tests is to verify if non of the CPU flags hasn't changed during normal syscalls and "LibOS syscalls".

## Useful links for reviewers
[System V Application Binary Interface - AMD64 Architecture Processor Supplement](https://refspecs.linuxbase.org/elf/x86_64-abi-0.99.pdf)

## Useful places in code
[Initialization of auxiliary vector](https://github.com/gramineproject/gramine/blob/2628ef6ba7df94fcdee811641d72bf2b835ca5d9/LibOS/shim/src/shim_rtld.c#L875)

## How to test this PR? 
```
gramine-test pytest
gramine-test --sgx pytest
```

---
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)

### CODE REVIEW

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @oshogbo)

___
*[`LibOS/shim/test/abi/x86_64/meson.build` line 20 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GB89j6K__YpnCpERp:-N1GB89j6K__YpnCpERq:bl91xd6) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/meson.build#L20)):*
> ```Meson
>         ],
>         'include_directories': [
>             # For 'elf/elf.h'
> ```

Wut? Why can't you include host `elf.h`?
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 14 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GCMvVD0oiEVGdZk5h:-N1GCMvVD0oiEVGdZk5i:b-jo4a4d) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L14)):*
> ```C
> int verify_auxiliary(Elf64_auxv_t *auxv);
> 
> struct {
> ```

NP `static`.
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 16 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GBPXZ7_0JU21gmM_Z:-N1GBPXZ7_0JU21gmM__:b-hly1re) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L16)):*
> ```C
> struct {
>     uint64_t type;
>     int exists;
> ```

`bool exists`?
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 30 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GCWRT2b6tPXMARhEP:-N1GCWRT2b6tPXMARhEQ:b-xocqgj) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L30)):*
> ```C
> int verify_auxiliary(Elf64_auxv_t *auxv) {
>     uint64_t count = sizeof(auxv_gramine_defaults) / sizeof(auxv_gramine_defaults[0]);
>     uint64_t i;
> ```

1. You can define this inside `for`.
2. We use `size_t` for iterators (basically just a convention).
___
*[`LibOS/shim/test/abi/x86_64/test_entrypoint.py` line 27 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GDhpN5GmWnR7SiiJ8:-N1GDhpN5GmWnR7SiiJ9:b-cb2kfa) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/test_entrypoint.py#L27)):*
> ```Python
>         self.run_binary(['stack_env'])
> 
>     def test_000_auxv(self):
> ```

Wait, why all of these have `000`?

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: all files reviewed, 12 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @boryspoplawski and @oshogbo)

___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HsKq_4BmeGPQK3TOn):*
This phrase sounds a bit weird. Maybe rephrase to `so the envvar tests come with a manifest file to set up envvars`.
___
*[`-- commits` line 24 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1Hso_t3pp64B8SeEdI):*
Why is this sentence on a separate line?
___
*[`LibOS/shim/test/abi/x86_64/meson.build` line 15 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1Ht4ly1EZYtiM9QJLQ:-N1Ht4ly1EZYtiM9QJLR:b-hbvupz) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/meson.build#L15)):*
> ```Meson
>     'stack_env': {},
>     'stack_auxiliary': {
>         'filename': [
> ```

Maybe `filenames` (plural)?
___
*[`LibOS/shim/test/abi/x86_64/meson.build` line 20 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GB89j6K__YpnCpERp:-N1Hst-pDNojsKPnA5gH:b2kf7rz) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/meson.build#L20)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Wut? Why can't you include host `elf.h`?
</blockquote></details>

+1 to Borys.
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 12 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HtQ0xBXiiSGGECIff:-N1HtQ0xBXiiSGGECIfg:b-mu9618) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L12)):*
> ```C
>  * This function is used with NASM, so creating a header is pointless.
>  */
> int verify_auxiliary(Elf64_auxv_t *auxv);
> ```

We have the asterisk aligned to the left, so `Elf64_auxv_t* auxv`
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 26 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HtaavA_hy2N5Ko8hc:-N1HtaavA_hy2N5Ko8hd:b-43h11l) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L26)):*
> ```C
>     { AT_PHENT, 0 },
>     { AT_SYSINFO_EHDR, 0 },
> };
> ```

Maybe add a link / path to the relevant Gramine code: `gramine/LibOS/shim/src/shim_rtld.c: execute_elf_object()`?
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 28 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HuFgIAYg_flEqDpzG:-N1HuFgIAYg_flEqDpzH:b-7ropln) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L28)):*
> ```C
> };
> 
> int verify_auxiliary(Elf64_auxv_t *auxv) {
> ```

ditto
___
*[`LibOS/shim/test/abi/x86_64/stack_env.asm` line 8 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HvjjaA0vYd38bSt4X:-N1HvjjaA0vYd38bSt4Y:b-5rq5fc) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_env.asm#L8)):*
> ```m68k
> ; In this test we don't pass any additional arguments so argc is 1.
> 
> default   rel
> ```

Why you have relative addressing here but not in the other asm file?
___
*[`LibOS/shim/test/abi/x86_64/test_entrypoint.py` line 27 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GDhpN5GmWnR7SiiJ8:-N1HuyLq7LyofW15aNj3:bl50byg) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/test_entrypoint.py#L27)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Wait, why all of these have `000`? 
</blockquote></details>

Oops, right, the tests should have some numbering. We do it pretty arbitrarily, with a step of 10 or 100, trying to group into topics.

Basically, please just do some numbering on the tests. I suggest with step 10 (so in the future, some tests can be plugged in-between).

**Review by oshogbo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: 3 of 7 files reviewed, 12 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "squash! " found in commit messages' one-liners (waiting on @boryspoplawski and @dimakuv)

___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HsKq_4BmeGPQK3TOn):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

This phrase sounds a bit weird. Maybe rephrase to `so the envvar tests come with a manifest file to set up envvars`.
</blockquote></details>

Done.
___
*[`-- commits` line 24 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1Hso_t3pp64B8SeEdI):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Why is this sentence on a separate line?
</blockquote></details>

Done.
___
*[`LibOS/shim/test/abi/x86_64/meson.build` line 15 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1Ht4ly1EZYtiM9QJLQ:-N1IrOHv2-6bzHxQ4DES:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/meson.build#L15)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Maybe `filenames` (plural)?
</blockquote></details>

Done.
___
*[`LibOS/shim/test/abi/x86_64/meson.build` line 20 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GB89j6K__YpnCpERp:-N1Iqwle9URvEhN8qde_:b-3l2qo2) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/meson.build#L20)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

+1 to Borys.
</blockquote></details>

We don't want to use elf/elf.h shipped with libPal?
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 12 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HtQ0xBXiiSGGECIff:-N1IrgTQ1BuUkKiN2XQl:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L12)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

We have the asterisk aligned to the left, so `Elf64_auxv_t* auxv`
</blockquote></details>

Done.
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 14 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GCMvVD0oiEVGdZk5h:-N1Irgsg-qchxIB0Et81:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L14)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

NP `static`.
</blockquote></details>

Done.
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 16 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GBPXZ7_0JU21gmM_Z:-N1IrnCpE6w6uGd5S4wI:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L16)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

`bool exists`?
</blockquote></details>

Done.
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 26 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HtaavA_hy2N5Ko8hc:-N1IrwlL9NkRElH4g93n:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L26)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Maybe add a link / path to the relevant Gramine code: `gramine/LibOS/shim/src/shim_rtld.c: execute_elf_object()`?
</blockquote></details>

Done.
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 28 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HuFgIAYg_flEqDpzG:-N1IrxAGDop6sWsFn6XY:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L28)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 30 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GCWRT2b6tPXMARhEP:-N1Is1g58wlLq4fknxct:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L30)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

1. You can define this inside `for`.
2. We use `size_t` for iterators (basically just a convention).
</blockquote></details>

Done.
___
*[`LibOS/shim/test/abi/x86_64/stack_env.asm` line 8 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HvjjaA0vYd38bSt4X:-N1IsQu76QhVNIp144Ye:b-s3hmk9) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_env.asm#L8)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Why you have relative addressing here but not in the other asm file?
</blockquote></details>

Not needed.
___
*[`LibOS/shim/test/abi/x86_64/test_entrypoint.py` line 27 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GDhpN5GmWnR7SiiJ8:-N1ItHpWASQOPC5RGJUm:b-8qi6so) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/test_entrypoint.py#L27)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Oops, right, the tests should have some numbering. We do it pretty arbitrarily, with a step of 10 or 100, trying to group into topics.

Basically, please just do some numbering on the tests. I suggest with step 10 (so in the future, some tests can be plugged in-between).
</blockquote></details>

I assumed its for numbering the same group of tests but with different args or so.
Fixed.

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "squash! " found in commit messages' one-liners (waiting on @boryspoplawski and @oshogbo)

___
*[`LibOS/shim/test/abi/x86_64/meson.build` line 20 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GB89j6K__YpnCpERp:-N1J41SK864wpdx3w8EE:b-5fx2gc) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/meson.build#L20)):*
<details><summary><i>Previously, oshogbo (Mariusz Zaborski) wrote…</i></summary><blockquote>

We don't want to use elf/elf.h shipped with libPal?
</blockquote></details>

We don't want to use this Gramine-internal file for Gramine-unrelated tests (I mean, these tests are just normal Linux binaries). So I don't see any need to use Gramine's header file `elf.h`.
___
*[`LibOS/shim/test/abi/x86_64/meson.build` line 30 at r2](https://reviewable.io/reviews/gramineproject/gramine/565#-N1J4Miw5eXtJbCkOdNx:-N1J4Miw5eXtJbCkOdNy:b-1zvor) ([raw file](https://github.com/gramineproject/gramine/blob/b83b23bffb814724e1b93e0fc3337b46d0410f68/LibOS/shim/test/abi/x86_64/meson.build#L30)):*
> ```Meson
> 
> foreach name, params : tests
>     filename = ''
> ```

I expected you to change this to `filenames` as well

**Review by dimakuv — APPROVED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @boryspoplawski)

**Comment by dimakuv** [comment]
Jenkins, test this please

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> Reviewed 2 of 4 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel), "fixup! " found in commit messages' one-liners (waiting on @oshogbo)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/565#-N1JKWPABezOUIrm54bh:-N1JKWPABezOUIrm54bi:bfi5d2p) (no related file):*
Blocking until https://github.com/gramineproject/gramine/pull/567 is merged

---
___
*[`LibOS/shim/test/abi/x86_64/meson.build` line 20 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1GB89j6K__YpnCpERp:-N1JKjuW5z5H98OwDMuG:b6wo8j0) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/meson.build#L20)):*
<details><summary><i>Previously, oshogbo (Mariusz Zaborski) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Yes, this has nothing to do with Gramine internals. These tests should also pass on host Linux.
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 9 at r3](https://reviewable.io/reviews/gramineproject/gramine/565#-N1JLEFuB4V3iwNRAJPJ:-N1JLEFuB4V3iwNRAJPK:b-rlcsvl) ([raw file](https://github.com/gramineproject/gramine/blob/bee0906b4fe493e1bf47c18453556e8d0025856d/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L9)):*
> ```C
> #include <stddef.h>
> 
> #include <elf.h>
> ```

Please sort
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 31 at r3](https://reviewable.io/reviews/gramineproject/gramine/565#-N1JLSTQ0r8V2kRHlTf-:-N1JLSTQ0r8V2kRHlTf0:bltarpk) ([raw file](https://github.com/gramineproject/gramine/blob/bee0906b4fe493e1bf47c18453556e8d0025856d/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L31)):*
> ```C
>     { AT_RANDOM, 0 },
>     { AT_PHENT, 0 },
>     { AT_SYSINFO_EHDR, 0 },
> ```

These all should have `false` now :)
___
*[`LibOS/shim/test/abi/x86_64/stack_env.asm` line 8 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HvjjaA0vYd38bSt4X:-N1JL_za3_7Wc5xK8dng:b-7qp3zj) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_env.asm#L8)):*
<details><summary><i>Previously, oshogbo (Mariusz Zaborski) wrote…</i></summary><blockquote>

Not needed.
</blockquote></details>

Wait, won't `lea rdi, [env0]` now produce a relocation?

**Review by oshogbo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: 6 of 7 files reviewed, 4 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @boryspoplawski, @dimakuv, and @oshogbo)

___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 9 at r3](https://reviewable.io/reviews/gramineproject/gramine/565#-N1JLEFuB4V3iwNRAJPJ:-N1KrSifCUy6RhaHb6qI:b-iqnxhc) ([raw file](https://github.com/gramineproject/gramine/blob/bee0906b4fe493e1bf47c18453556e8d0025856d/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L9)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Please sort
</blockquote></details>

I wanted to separate them as a standard group and others, but I will merge them.
___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 31 at r3](https://reviewable.io/reviews/gramineproject/gramine/565#-N1JLSTQ0r8V2kRHlTf-:-N1KrRCcB7MV5rT7byD_:b-80tlqk) ([raw file](https://github.com/gramineproject/gramine/blob/bee0906b4fe493e1bf47c18453556e8d0025856d/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L31)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

These all should have `false` now :)
</blockquote></details>

Sure.
___
*[`LibOS/shim/test/abi/x86_64/stack_env.asm` line 8 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HvjjaA0vYd38bSt4X:-N1KsjXv6Z-rjN0hIMrJ:bok3fl5) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_env.asm#L8)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Wait, won't `lea rdi, [env0]` now produce a relocation?
</blockquote></details>

Yes, so in other tests we don't have a data segment in assembly. The `stack_arg` and `stack_env` has it and I used a relative addressing here.
In case of NASM it just the different if we are accessing data through `rsp` or `ds` register. 

Actually it doesn't create a relocation entry:
```
$ readelf -r stack_env_without_rel
There are no relocations in this file.

$ readelf -r stack_env_rel
There are no relocations in this file.
```

Or I miss understand something.



_Code quote:_
```m68k
default   rel
```

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @oshogbo)

___
*[`LibOS/shim/test/abi/x86_64/stack_auxiliary.c` line 9 at r3](https://reviewable.io/reviews/gramineproject/gramine/565#-N1JLEFuB4V3iwNRAJPJ:-N1L2TndAKPCJ1cX8u1U:b-ceye0a) ([raw file](https://github.com/gramineproject/gramine/blob/bee0906b4fe493e1bf47c18453556e8d0025856d/LibOS/shim/test/abi/x86_64/stack_auxiliary.c#L9)):*
<details><summary><i>Previously, oshogbo (Mariusz Zaborski) wrote…</i></summary><blockquote>

I wanted to separate them as a standard group and others, but I will merge them.
</blockquote></details>

This is a standard header...
___
*[`LibOS/shim/test/abi/x86_64/stack_env.asm` line 8 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HvjjaA0vYd38bSt4X:-N1L6BWvDp2Ugoqd-qBs:b-p1qzza) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_env.asm#L8)):*
<details><summary><i>Previously, oshogbo (Mariusz Zaborski) wrote…</i></summary><blockquote>

Yes, so in other tests we don't have a data segment in assembly. The `stack_arg` and `stack_env` has it and I used a relative addressing here.
In case of NASM it just the different if we are accessing data through `rsp` or `ds` register. 

Actually it doesn't create a relocation entry:
```
$ readelf -r stack_env_without_rel
There are no relocations in this file.

$ readelf -r stack_env_rel
There are no relocations in this file.
```

Or I miss understand something.
</blockquote></details>

There is no real ds segment on x64.
It doesn't create relocations because the binary is non-pie, so it knows addresses in advance. But it would create them, if it was pie.
With `default rel`:
```
401013:       48 8d 3d e6 2f 00 00    lea    rdi,[rip+0x2fe6]        # 404000 <argv0>
```
without:
```
401013:       48 8d 3c 25 00 40 40    lea    rdi,ds:0x404000
```
So either please add it, or use `rel` manually.

**Review by mkow — APPROVED** [review]
> Reviewed 3 of 7 files at r1, 2 of 4 files at r2, 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @oshogbo)

___
*[`-- commits` line 37 at r4](https://reviewable.io/reviews/gramineproject/gramine/565#-N1LkZXiDhYYxNxc-WU2):*
Please wrap the document title in doublequotes, will be easier to see where the title ends (or we can do it ourselves at the final rebase).

**Review by dimakuv — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners (waiting on @boryspoplawski and @oshogbo)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/565#-N1JKWPABezOUIrm54bh:-N1N0X9Z3upRhChiMrCx:b2fay) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Blocking until https://github.com/gramineproject/gramine/pull/567 is merged
</blockquote></details>

#567 is merged

---

**Review by oshogbo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: 6 of 7 files reviewed, 3 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "squash! " found in commit messages' one-liners (waiting on @boryspoplawski, @dimakuv, @mkow, and @oshogbo)

___
*[`LibOS/shim/test/abi/x86_64/stack_env.asm` line 8 at r1](https://reviewable.io/reviews/gramineproject/gramine/565#-N1HvjjaA0vYd38bSt4X:-N1NHmoW3rQSP1ZB3DT9:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/b8c02e6256c9dc7437ffc2305863b913a227835c/LibOS/shim/test/abi/x86_64/stack_env.asm#L8)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

There is no real ds segment on x64.
It doesn't create relocations because the binary is non-pie, so it knows addresses in advance. But it would create them, if it was pie.
With `default rel`:
```
401013:       48 8d 3d e6 2f 00 00    lea    rdi,[rip+0x2fe6]        # 404000 <argv0>
```
without:
```
401013:       48 8d 3c 25 00 40 40    lea    rdi,ds:0x404000
```
So either please add it, or use `rel` manually.
</blockquote></details>

Done.

**Review by mkow — DISMISSED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "squash! " found in commit messages' one-liners (waiting on @boryspoplawski and @oshogbo)

**Review by boryspoplawski — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "squash! " found in commit messages' one-liners (waiting on @oshogbo)

___
*[`-- commits` line 70 at r5](https://reviewable.io/reviews/gramineproject/gramine/565#-N1P27Vc5i6wYXPzbtAS):*
Did you mean to put something in this `squash` or?

**Comment by boryspoplawski** [comment]
Jenkins, test this please

**Comment by dimakuv** [comment]
@oshogbo Can you squash and rebase your PR to the latest master? It is ready for merge.

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @mkow)

___
*[`-- commits` line 37 at r4](https://reviewable.io/reviews/gramineproject/gramine/565#-N1LkZXiDhYYxNxc-WU2):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Please wrap the document title in doublequotes, will be easier to see where the title ends (or we can do it ourselves at the final rebase).
</blockquote></details>

I've done it.

**Comment by dimakuv** [comment]
Jenkins, test this please

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/565)* status: :shipit: complete! all files reviewed, all discussions resolved
